provider/gce/google/errors.go

```diff
@@ -120,3 +120,11 @@ var AuthorisationFailureStatusCodes = map[int][]string{
 	// https://tools.ietf.org/html/rfc6749#section-5.2
 	http.StatusBadRequest: {"Bad Request"},
 }
+
+// IsNotFound reports whether err contains `not found'.
+func IsNotFound(err error) bool {
```

Comment:
From memory the best way to do this is:
```
if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == http.StatusNotFound {
}
```

Reply:
Yes, that covers some errors although ```errors.Cause(err).(*googleapi.Error)``` is a better.

However, google provider is notoriously inconsistent with errors and the above does not cover for example ```juju.provider.gce.gceapi raw.go:368 GCE operation error: (RESOURCE_NOT_FOUND) The resource 'projects/gothic-list-89514/zones/us-east1-b/instances/juju-f22d51-8' was not found```

I am trying to test with this + examining the gerr.Error() but so far the quickest check that covers all cases is my original one. In fact, since errors.IsNotFound is essentially checking for 'not found'  suffix as well, my original code covers that case too.

Reply:
errors.IsNotFound isn't checking for "not found", it looks for a specific type. I'd rather the check here looked for 3 specific types (if it turns out that RemoveInstance, RemoveFirewall and GetFirewalls all return different ones), rather than checking the string value for "not found". Using the string value seems both overly broad (might match unexpected errors) and fragile - it'll fail if they change the message to something more user-friendly like "couldn't be found" while the type stays the same.

Reply:
Ah, it's *compute.OperationErrorErrors... I'll see if I can add it as a 3rd check

Reply:
and of course compute.OperationErrorErrors does not implement error :( so i'd still need to check for "not found" string in this case...

At this stage, I do not see the value of type asserting. It's easier to just have a check for 'not found' sting than errors.NotFound, googleapi.Error and still check the string.

Reply:
Intriguingly, looking at the code I don't think the actual error that will be returned in that case includes the message "not found". 

The error's being logged from provider/gce/google/raw.go:368, but the error that will be returned is just `waitError{op, nil}`, so the not found error will have been shown to the user and then we'll show a generic "GCE operation blah failed" because it won't match the "not found" check higher up. I think the best way to fix this would be to assume that in the vast majority of cases op.Error.Errors will have 1 error in it, so pull out the first one, log any others and return the first one in the waitError. Then add a .Cause method to waitError so you can pull the internal error out with errors.Cause for your not found check higher up.

What do you think?

Reply:
Yes, so I have decided to handle these operational compute 'errors'. There is a func that preserves current behavior for majority of cases, i.e. just log these errors. There is another one to inspect them and if any contain the code I am encountering, then actually convert non-error struct into an error we all recognise and love.
I thin it's a cleaner approach. Have a look - this change is in raw.go. It also works a treat - just tested it live.

We can expand on what we do with different errors or add to the list we consider 'not found' as the need arises.